### DISCUSSION TITLE
javascript do not work after navigating away in SPA

### DISCUSSION THREAD

**Discussion by narutodiablo** [livewire/livewire#7771] [1. Help]
I have a big problem with my Livewire 3 project, I am using livewire 3 without vite/tailwind just plain livewire 3 and bootstrap 5. In one of my SPA dashboards, i have a fullcalendar plugin, but when i navigate into another component and go back to the dashboard within the SPA, the fullcalendar disappears. 

i have seen the flowbite and tried incorporating it in my /resources/js/app.js  however its still not working

```
import './bootstrap';

document.addEventListener('livewire:navigated', ()=>{
    initFlowbite();
});
```

This is how i imported it in my layout page <script src="https://cdnjs.cloudflare.com/ajax/libs/flowbite/2.2.1/flowbite.min.js"></script> but it still wont work.

how can i fix this so that my javascript components (fullcalendar, select2, chartjs) would still work when i navigate away from the component and return to it from within SPA?

**Comment by danie-ramdhani** [comment]
did you import your script in body tag?
then you need to add `data-navigate-once` attribute to your script (https://livewire.laravel.com/docs/navigate#scripts-in-the-body-are-re-evaluated)

```html
<script src="https://cdnjs.cloudflare.com/ajax/libs/flowbite/2.2.1/flowbite.min.js" data-navigate-once></script>
```

**Reply by narutodiablo** [reply]
thanks for the response. yes I did put it inside my body tag before the closing </body> tag. and i also tried what you suggested, i added data-navigate-once but it still didnt fix the issue. 

i experimented as well in another SPA i have where i was toggling the state of the component and using if statements to show or hide the components within the page but its the same problem.

**Comment by narutodiablo** [comment]
is there any solution to this issue?

**Comment by shubdragon** [comment]
can you do this in the bottom of your body tag

```
<body>
//blahh..
<script type="module">
document.addEventListener('livewire:navigated', ()=>{
    initFlowbite();
});
</script>
</body
```

It is because possibly you have initialized flowbite and its asets in head using `@vite([blahhh])` but on  navigated is not initialized for new elements which was dynamically replaced by livewire, so if you call initFLowbite again on navigated it will work also add data-navigate-once to prevent multiple call

**Reply by narutodiablo** [reply]
i did but still the same issue.

**Reply by Omar7tech** [reply]
### Thank You! It Worked Perfectly! 🎉  
Here’s the solution that worked for me:  

```html
<script type="module">
    document.addEventListener('livewire:navigated', () => {
        initFlowbite();
    });
</script>
```  

This simple script ensures `initFlowbite()` runs after Livewire navigation, resolving any issues I had.  

---

### If This Doesn’t Work for You, Try These Steps:  

1. **Update Livewire**  
   Make sure you’re using the latest version of Livewire. You can do this by running:  
   ```bash
   composer update livewire/livewire
   ```  

2. **Clear Cache**  
   If the update doesn’t resolve your issue, clearing the cache might help. Use this command:  
   ```bash
   php artisan optimize:clear
   ```  

---

Hope this helps others too! 😊 If you have any questions or need further clarification, feel free to ask. 🚀


```html
<!DOCTYPE html>
<html lang="{{ str_replace('_', '-', app()->getLocale()) }}">
<head>
    <meta charset="utf-8">
    <meta name="viewport" content="width=device-width, initial-scale=1.0">
    <title>{{ $title ?? 'SPA Livewire Train' }}</title>
    @vite(['resources/css/app.css', 'resources/js/app.js'])
</head>
<body class="dark:bg-gray-900">

    <x-nav.main>
        {{ $slot }}
    </x-nav.main>

    {{--* PLACE THE SCRIPT HERE --}}
    <script type="module">
        document.addEventListener('livewire:navigated', () => {
            initFlowbite();
        });
    </script>
    
</body>
</html>

```

**Comment by narutodiablo** [comment]
i would like to add more details on what i am doing. an example is i have 3 components

```
php artisan make:livewire Auth/Main
php artisan make:livewire Auth/Login
php artisan make:livewire Auth/Register
```

what i do is i load the Auth/Main component into my blade that i then toggle the states between login and register components

app/Http/Livewire/Auth/Main.php

```
namespace App\Http\Livewire\Auth;

use Livewire\Component;

class Main extends Component
{
    public $showLogin = true; // Show login form by default

    public function toggleForm()
    {
        $this->showLogin = !$this->showLogin; // Toggle between login and register
    }

    public function render()
    {
        return view('livewire.auth.main');
    }
}
```

resources/views/livewire/auth/main.blade.php

```
<div>
    @if ($showLogin)
        @livewire('auth.login')
    @else
        @livewire('auth.register')
    @endif

    <button class="btn btn-link" wire:click="toggleForm()">
        @if ($showLogin)
            Don't have an account? Register
        @else
            Already have an account? Login
        @endif
    </button>
</div>
```

With this i can toggle between components, however, if for example i have a select2 javascript on my field or i have a special javascript form field, when i go to login and go back to register component, the javascripts do not work anymore.

please help me fix this or if there is a better way of handling this.

**Reply by shubdragon** [reply]
okz, that means it is actually not navigated just including livewire component based on condition.

okay here one thing you can do on your  base view or app.blade.php add this script:
```
<script>
    window.addEventListener('load-my-script', function(){
        initFlowbite();
        console.log('hi');
    })
</script>

```

Then on your toggle method

```
public function toggleForm()
{       
        $this->showLogin = !$this->showLogin; // Toggle between login and register
        $this->dispatch('load-my-script');
}
```

**Reply by narutodiablo** [reply]
thanks for the response, unfortunately its still not working.

```
<script>
    window.addEventListener('livewire-reload',  function(){
        initFlowbite();
        console.log('reload-fired');
    });
</script>
```

when i do this, i see the console logs as intended (reload fired) and no errors are being shown but still the javascript libraries do not reload.  i tried this on my pages that load fullcalendar and charts and still the same.

i have this embedded in my header

` <script src="https://cdnjs.cloudflare.com/ajax/libs/flowbite/2.2.1/flowbite.min.js" data-navigate-once></script>`

**Reply by shubdragon** [reply]
never used flowbite before, 
Is fullcalendar and chart js also initialize on initFlowbite(); ? if not you have to initialize fullcalendar and chart js both manually from that script tag similar to console log

what happens if you manually execute initFlowbite(); from browser console window ? will it  work or not.

you can also try to install it with npm so the @vite([blahh]) already includes its js for you and then you need to initialize flowbite on each include, then above solution will work

(*assuming flowbite js not working only for newly included livewire component using if condition)

Also try to include the cdn on head tag instead so it wont be replaced

**Reply by narutodiablo** [reply]
thanks again for the response, i think i am starting to get what you mean. lets disregard flowbite for now, for instance, i have the fullcalendar in my component blade:

**@script**
`<script>

    let calendarEl = document.getElementById('workorder-calendar');

    let calendar = new FullCalendar.Calendar(calendarEl, {
        // FullCalendar's options and event sources
        timeZone: 'Asia/Seoul',
        themeSystem: 'bootstrap5',
        headerToolbar: {
            right: 'prev today next',
        },
        footerToolbar: {
            left: 'dayGridMonth,timeGridWeek,timeGridDay,listMonth'
        },
        weekNumbers: true,
        dayMaxEvents: true,
    });

    calendar.render();

</script>`
**@endscript**

how can i call the javascript in the <script> tag so that they can be reinitialized when i get back to this component using the dispatch method?

**Reply by shubdragon** [reply]
okay if you use alpine js i have more elegant solution else you could try this: (because i will prefer to make a portable component using alpine js, so it could be reuse)


on base page:


```
@script
<script>



function initCalendar(){
let calendarEl = document.getElementById('workorder-calendar');
let calendar = new FullCalendar.Calendar(calendarEl, {
    // FullCalendar's options and event sources
    //let calendar again because  element was actually replaced and it is just a reference of old element.
    timeZone: 'Asia/Seoul',
    themeSystem: 'bootstrap5',
    headerToolbar: {
        right: 'prev today next',
    },
    footerToolbar: {
        left: 'dayGridMonth,timeGridWeek,timeGridDay,listMonth'
    },
    weekNumbers: true,
    dayMaxEvents: true,
});

calendar.render();
}


initCalendar();
</script>
@endscript

```
now on toggle method of your alpine dispatch the  livewire-reload

and your js:
```
<script>
    window.addEventListener('livewire-reload',  function(){
        initFlowbite();
        initCalendar();
        console.log('reload-fired');
    });
</script>
```

**If use alpinejs don't need to do all this alpine init does it for you on each page

**Reply by narutodiablo** [reply]
thanks, i tried this but i now get the error that says initCalendar is not defined

`
Alpine Expression Error: initCalendar is not defined 
`

i would like to clarify that the initCalendar is inside my component blade, and the livewire-reload eventlister is in my base layout file.

**Reply by shubdragon** [reply]
Yes I thought you will get this error next, just it should be from general javascript, if you are using alpine then it could be handle differently inline,

because  on initial page load this function may be available but on wire navigated it must be lost.


if you are using alpine.js then directly add this div where you need the calendar or alternately make it as a component, and it is also need to be noted if you make it as a component and will have multiple calendar in single page don't forget to make datePickerInstance unique by attching id or something as name e.g ` x-data="datePickerInstance_{{$attributes->get('id')}}" ` something

```
<div  wire:ignore x-data="{datePickerInstance: null}" 
x-init="datePickerInstance =  new FullCalendar.Calendar($el, {    
    timeZone: 'Asia/Seoul',
    themeSystem: 'bootstrap5',
    headerToolbar: {
        right: 'prev today next',
    },
    footerToolbar: {
        left: 'dayGridMonth,timeGridWeek,timeGridDay,listMonth'
    },
    weekNumbers: true,
    dayMaxEvents: true,
}); datePickerInstance.render();"></div>
```
// not tried just typed here check if it works for you, don't forget to reply if worked
// don't need to dispatch anything now
